src/workflow_provider/default_vm_workflow_provider.cpp

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (C) 2021 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <multipass/default_vm_workflow_provider.h>
+#include <multipass/exceptions/download_exception.h>
+#include <multipass/exceptions/invalid_memory_size_exception.h>
+#include <multipass/exceptions/workflow_exceptions.h>
+#include <multipass/format.h>
+#include <multipass/logging/log.h>
+#include <multipass/query.h>
+#include <multipass/url_downloader.h>
+#include <multipass/utils.h>
+
+#include <yaml-cpp/yaml.h>
+
+#include <QFile>
+#include <QFileInfo>
+
+#include <Poco/StreamCopier.h>
+#include <Poco/Zip/ZipArchive.h>
+#include <Poco/Zip/ZipStream.h>
+
+#include <fstream>
+#include <sstream>
+
+namespace mp = multipass;
+namespace mpl = multipass::logging;
+
+namespace
+{
+const QString github_workflows_archive_name{"multipass-workflows.zip"};
+const QString workflow_dir_version{"v1"};
+constexpr auto category = "workflow provider";
+
+auto workflows_map_for(const std::string& archive_file_path)
+{
+    std::map<std::string, std::string> workflows_map;
+    std::ifstream zip_stream{archive_file_path, std::ios::binary};
+    Poco::Zip::ZipArchive zip_archive{zip_stream};
+
+    for (auto it = zip_archive.headerBegin(); it != zip_archive.headerEnd(); ++it)
```

Comment:
Should we do any error handling here to see if the zip is bad? There's a [`Poco::Zip::ZipException`](https://pocoproject.org/docs/Poco.Zip.ZipException.html)?

Reply:
Yeah, I think it would be prudent to do that.

Reply:
I will note that POCO is quite heavy handed on how it handles some error conditions and will call `std::abort()` if an assertion fails.  There is really no way for us to recover from such a case either.  For example, say a zip file is corrupted and is missing some expected info in the zip archive, POCO will fail an assertion and call `std::abort()` and we will bail.  Can't say I really like that approach from POCO as I think it should throw an exception instead and let the consumer of the library deal with it as it sees fit.

Reply:
Here's an example in POCO about what I'm describing above:
https://github.com/pocoproject/poco/blob/3fc3e5f5b8462f7666952b43381383a79b8b5d92/Zip/src/ZipArchive.cpp#L78
and this is what `poco_assert()` does:
https://github.com/pocoproject/poco/blob/3fc3e5f5b8462f7666952b43381383a79b8b5d92/Foundation/include/Poco/Bugcheck.h#L95

Reply:
Scratch the last two things I posted.  The assertions with `std::abort()` shouldn't be too much of a problem.  I do see where it should throw an exception when the zip is not correct.  However, I'm directly modifying a zip file to put in a bad header and POCO doesn't seem to notice, so I'll look more into that.

Reply:
Ok, doing some debugging in POCO, it has a bug in how it handles bad headers, etc. in a zip file.  I'll probably see if I can fix that sometime and upstream it.  I'll also see if there is an existing bug and if not, enter one in the POCO project.

That said, there won't be a good way to test zip exceptions other than put the zip stuff under control of a singleton and inject mocks to simulate exceptions.  I'll see about that.

Reply:
This should be addressed now.